File: app/src/main/java/igrek/todotree/ui/treelist/TreeListReorder.java
package igrek.todotree.ui.treelist;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.ObjectAnimator;
import android.animation.TypeEvaluator;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.drawable.BitmapDrawable;
import android.view.View;

import java.util.List;

import igrek.todotree.domain.treeitem.AbstractTreeItem;
import igrek.todotree.info.logger.Logger;
import igrek.todotree.info.logger.LoggerFactory;
import igrek.todotree.intent.TreeCommand;

import static android.view.View.INVISIBLE;
import static android.view.View.VISIBLE;

public class TreeListReorder {
	
	private Logger logger = LoggerFactory.INSTANCE.getLogger();
	private TreeListView listView;
	
	/** początkowa pozycja kursora przy rozpoczęciu przeciągania (ulega zmianie przy zamianie elementów) */
	private float startTouchY;
	/** aktualna pozycja kursora względem ekranu i listview (bez scrollowania) */
	private float lastTouchY;
	/** położenie scrolla przy rozpoczęciu przeciągania */
	private Integer scrollStart = 0;
	
	/** pozycja aktualnie przeciąganego elementu */
	private Integer draggedItemPos = null;
	/** widok aktualnie przeciąganego elementu */
	private View draggedItemView = null;
	/** oryginalna górna pozycja niewidocznego elementu na liście, którego bitmapa jest przeciągana */
	private Integer draggedItemViewTop = null;
	
	private BitmapDrawable hoverBitmap = null;
	private BitmapDrawable hoverBitmapAnimation = null;
	private Rect hoverBitmapBounds;
	
	private final float ITEMS_REPLACE_COVER = 0.65f;
	
	private final int HOVER_BORDER_THICKNESS = 5;
	private final int HOVER_BORDER_COLOR = 0xccb0b0b0;
	
	public TreeListReorder(TreeListView listView) {
		this.listView = listView;
	}
	
	private void updateHoverBitmap() {
		if (draggedItemViewTop != null && draggedItemPos != null) {
			float dy = lastTouchY - startTouchY;
			hoverBitmapBounds.offsetTo(0, draggedItemViewTop + (int) dy);
			hoverBitmap.setBounds(hoverBitmapBounds);
		}
	}
	
	private BitmapDrawable getAndAddHoverView(View v) {
		int top = v.getTop();
		int left = v.getLeft();
		
		Bitmap b = getBitmapWithBorder(v);
		BitmapDrawable drawable = new BitmapDrawable(listView.getResources(), b);
		
		hoverBitmapBounds = new Rect(left, top, left + v.getWidth(), top + v.getHeight());
		drawable.setBounds(hoverBitmapBounds);
		
		return drawable;
	}
	
	private Bitmap getBitmapWithBorder(View v) {
		Bitmap bitmap = getBitmapFromView(v);
		Canvas can = new Canvas(bitmap);
		
		Rect rect = new Rect(0, 0, bitmap.getWidth(), bitmap.getHeight());
		
		Paint paint = new Paint();
		paint.setStyle(Paint.Style.STROKE);
		paint.setStrokeWidth(HOVER_BORDER_THICKNESS);
		paint.setColor(HOVER_BORDER_COLOR);
		
		can.drawBitmap(bitmap, 0, 0, null);
		can.drawRect(rect, paint);
		
		return bitmap;
	}
	
	private Bitmap getBitmapFromView(View v) {
		Bitmap bitmap = Bitmap.createBitmap(v.getWidth(), v.getHeight(), Bitmap.Config.ARGB_8888);
		Canvas canvas = new Canvas(bitmap);
		v.draw(canvas);
		return bitmap;
	}
	
	private final static TypeEvaluator<Rect> rectBoundsEvaluator = new TypeEvaluator<Rect>() {
		public Rect evaluate(float fraction, Rect startValue, Rect endValue) {
			return new Rect(interpolate(startValue.left, endValue.left, fraction), interpolate(startValue.top, endValue.top, fraction), interpolate(startValue.right, endValue.right, fraction), interpolate(startValue.bottom, endValue.bottom, fraction));
		}
		
		int interpolate(int start, int end, float fraction) {
			return (int) (start + fraction * (end - start));
		}
	};
	
	
	private void itemDraggingStarted(int position, View itemView) {
		draggedItemPos = position;
		draggedItemView = itemView;
		draggedItemViewTop = itemView.getTop();
		scrollStart = listView.getScrollHandler().getScrollOffset();
		
		hoverBitmap = getAndAddHoverView(draggedItemView);
		draggedItemView.setVisibility(INVISIBLE);
		
		listView.invalidate();
	}
	
	public void setLastTouchY(float lastTouchY) {
		this.lastTouchY = lastTouchY;
	}
	
	public void handleItemDragging() {
		
		float dyTotal = lastTouchY - startTouchY + (listView.getScrollHandler()
				.getScrollOffset() - scrollStart);
		
		if (draggedItemViewTop == null) {
			logger.error("draggedItemViewTop = null");
			return;
		}
		if (draggedItemPos == null) {
			logger.error("draggedItemPos = null");
			return;
		}
		
		int step = 0;
		int deltaH = 0;
		
		if (dyTotal > 0) { //przesuwanie w dół
			while (true) {
				if (draggedItemPos + step + 1 >= listView.getItems().size())
					break;
				int downHeight = listView.getItemHeight(draggedItemPos + step + 1);
				if (downHeight == 0)
					break;
				if (dyTotal - deltaH > downHeight * ITEMS_REPLACE_COVER) {
					step++;
					deltaH += downHeight;
				} else {
					break;
				}
			}
		} else if (dyTotal < 0) { //przesuwanie w górę
			while (true) {
				if (draggedItemPos + step - 1 < 0)
					break;
				int upHeight = listView.getItemHeight(draggedItemPos + step - 1);
				if (upHeight == 0)
					break;
				if (-dyTotal + deltaH > upHeight * ITEMS_REPLACE_COVER) {
					step--;
					deltaH -= upHeight;
				} else {
					break;
				}
			}
		}
		
		if (step != 0) {
			int targetPosition = draggedItemPos + step;
			
			List<AbstractTreeItem> items = new TreeCommand().itemMoved(draggedItemPos, step);
			
			//update mapy wysokości
			if (step > 0) {
				Integer draggedItemHeight = listView.getItemHeight(draggedItemPos);
				for (int i = draggedItemPos; i < targetPosition; i++) {
					Integer nextHeight = listView.getItemHeight(i + 1);
					listView.putItemHeight(i, nextHeight);
				}
				listView.putItemHeight(targetPosition, draggedItemHeight);
			} else if (step < 0) {
				Integer draggedItemHeight = listView.getItemHeight(draggedItemPos);
				for (int i = draggedItemPos; i > targetPosition; i--) {
					Integer nextHeight = listView.getItemHeight(i - 1);
					listView.putItemHeight(i, nextHeight);
				}
				listView.putItemHeight(targetPosition, draggedItemHeight);
			}
			
			listView.getAdapter().setDataSource(items);
			
			startTouchY += deltaH;
			draggedItemViewTop += deltaH;
			
			draggedItemPos = targetPosition;
			if (draggedItemView != null) {
				draggedItemView.setVisibility(VISIBLE);
			}
			draggedItemView = listView.getItemView(draggedItemPos);
			if (draggedItemView != null) {
				draggedItemView.setVisibility(INVISIBLE);
			}
		}
		
		listView.getScrollHandler().handleScrolling();
		
		listView.invalidate();
	}
	
	public void itemDraggingStopped() {
		if (draggedItemPos != null && draggedItemViewTop != null) {
			//wyłączenie automatycznego ustawiania pozycji hover bitmapy
			draggedItemPos = null;
			
			//animacja powrotu do aktualnego połozenia elementu
			// kopia referencji na czas trwania animacji
			hoverBitmapAnimation = hoverBitmap;
			Integer scrollOffset = listView.getScrollHandler().getScrollOffset();
			hoverBitmapBounds.offsetTo(0, draggedItemViewTop - (scrollOffset - scrollStart));
			final Rect hoverBitmapBoundsCopy = new Rect(hoverBitmapBounds);
			final View draggedItemViewCopy = draggedItemView;
			
			draggedItemView = null;
			hoverBitmap = null;
			draggedItemViewTop = null;
			
			ObjectAnimator hoverViewAnimator = ObjectAnimator.ofObject(hoverBitmapAnimation, "bounds", rectBoundsEvaluator, hoverBitmapBoundsCopy);
			hoverViewAnimator.addUpdateListener(valueAnimator -> listView.invalidate());
			hoverViewAnimator.addListener(new AnimatorListenerAdapter() {
				@Override
				public void onAnimationStart(Animator animation) {
					listView.setEnabled(false);
				}
				
				@Override
				public void onAnimationEnd(Animator animation) {
					if (draggedItemViewCopy != null) {
						draggedItemViewCopy.setVisibility(VISIBLE);
					}
					hoverBitmapAnimation = null;
					listView.setEnabled(true);
					listView.invalidate();
				}
			});
			hoverViewAnimator.start();
		} else {
			draggedItemPos = null;
			draggedItemView = null;
			draggedItemViewTop = null;
		}
	}
	
	public void onItemMoveButtonPressed(int position, AbstractTreeItem item, View itemView, float touchX, float touchY) {
		startTouchY = itemView.getTop() + touchY;
		lastTouchY = startTouchY;
		itemDraggingStarted(position, itemView);
	}
	
	public void onItemMoveButtonReleased(int position, AbstractTreeItem item, View itemView, float touchX, float touchY) {
		itemDraggingStopped();
	}
	
	public boolean isDragging() {
		return draggedItemPos != null;
	}
	
	public void setDraggedItemView() {
		View draggedItemViewOld = draggedItemView;
		draggedItemView = listView.getItemView(draggedItemPos);
		if (draggedItemView != null) {
			draggedItemView.setVisibility(View.INVISIBLE);
		}
		if (draggedItemViewOld != draggedItemView && draggedItemViewOld != null) {
			draggedItemViewOld.setVisibility(View.VISIBLE);
		}
	}
	
	public void dispatchDraw(Canvas canvas) {
		if (hoverBitmap != null) {
			updateHoverBitmap();
			hoverBitmap.draw(canvas);
		}
		if (hoverBitmapAnimation != null) {
			hoverBitmapAnimation.draw(canvas);
		}
	}
	
	public Rect getHoverBitmapBounds() {
		return hoverBitmapBounds;
	}
}

File: app/src/main/java/igrek/todotree/ui/treelist/TreeListReorder.kt
package igrek.todotree.ui.treelist

import android.animation.*
import android.graphics.*
import android.graphics.drawable.BitmapDrawable
import android.view.View
import igrek.todotree.domain.treeitem.AbstractTreeItem
import igrek.todotree.info.logger.LoggerFactory
import igrek.todotree.intent.TreeCommand


class TreeListReorder(private val listView: TreeListView) {

    private val logger = LoggerFactory.logger

    private var startTouchY = 0f
    private var lastTouchY = 0f
    private var scrollStart = 0
    private var draggedItemPos: Int? = null
    private var draggedItemView: View? = null
    private var draggedItemViewTop: Int? = null
    private var hoverBitmap: BitmapDrawable? = null
    private var hoverBitmapAnimation: BitmapDrawable? = null
    var hoverBitmapBounds: Rect? = null
        private set
    private val itemsReplaceCover = 0.65f
    private val hoverBorderThickness = 8
    private val hoverBorderColor = -0x334f4f50
    private val hoverTintColor: Long = 0x44aaaaaa
    private val treeCommand = TreeCommand()

    private fun updateHoverBitmap() {
        if (draggedItemViewTop != null && draggedItemPos != null) {
            val dy = lastTouchY - startTouchY
            hoverBitmapBounds!!.offsetTo(0, draggedItemViewTop!! + dy.toInt())
            hoverBitmap!!.bounds = hoverBitmapBounds!!
        }
    }

    private fun getAndAddHoverView(v: View?): BitmapDrawable {
        val top = v!!.top
        val left = v.left
        val b = getBitmapWithBorder(v)
        val drawable = BitmapDrawable(listView.resources, b)
        hoverBitmapBounds = Rect(left, top, left + v.width, top + v.height)
        drawable.bounds = hoverBitmapBounds!!
        return drawable
    }

    private fun getBitmapWithBorder(v: View?): Bitmap {
        val bitmap = getBitmapFromView(v)
        val canvas = Canvas(bitmap)

        val paint2 = Paint()
        paint2.colorFilter = PorterDuffColorFilter(getColor(hoverTintColor), PorterDuff.Mode.ADD)
        canvas.drawBitmap(bitmap, 0f, 0f, paint2)

        val paint = Paint()
        paint.style = Paint.Style.STROKE
        paint.strokeWidth = hoverBorderThickness.toFloat()
        paint.color = hoverBorderColor
        val rect = Rect(0, 0, bitmap.width, bitmap.height)
        canvas.drawRect(rect, paint)

        return bitmap
    }

    private fun getBitmapFromView(v: View?): Bitmap {
        val bitmap = Bitmap.createBitmap(v!!.width, v.height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        v.draw(canvas)
        return bitmap
    }

    private fun itemDraggingStarted(position: Int, itemView: View) {
        draggedItemPos = position
        draggedItemView = itemView
        draggedItemViewTop = itemView.top
        scrollStart = listView.scrollHandler!!.scrollOffset
        hoverBitmap = getAndAddHoverView(draggedItemView)
        draggedItemView!!.visibility = View.INVISIBLE
        listView.invalidate()
    }

    fun setLastTouchY(lastTouchY: Float) {
        this.lastTouchY = lastTouchY
    }

    fun handleItemDragging() {
        val dyTotal = lastTouchY - startTouchY + (listView.scrollHandler!!.scrollOffset - scrollStart)
        if (draggedItemViewTop == null) {
            logger.error("draggedItemViewTop = null")
            return
        }
        if (draggedItemPos == null) {
            logger.error("draggedItemPos = null")
            return
        }
        var step = 0
        var deltaH = 0
        if (dyTotal > 0) {
            while (true) {
                if (draggedItemPos!! + step + 1 >= listView.items!!.size) break
                val downHeight = listView.getItemHeight(draggedItemPos!! + step + 1)
                if (downHeight == 0) break
                deltaH += if (dyTotal - deltaH > downHeight * itemsReplaceCover) {
                    step++
                    downHeight
                } else {
                    break
                }
            }
        } else if (dyTotal < 0) {
            while (true) {
                if (draggedItemPos!! + step - 1 < 0) break
                val upHeight = listView.getItemHeight(draggedItemPos!! + step - 1)
                if (upHeight == 0) break
                deltaH -= if (-dyTotal + deltaH > upHeight * itemsReplaceCover) {
                    step--
                    upHeight
                } else {
                    break
                }
            }
        }
        if (step != 0) {
            val targetPosition = draggedItemPos!! + step
            val items = treeCommand.itemMoved(
                draggedItemPos!!, step
            )

            if (step > 0) {
                val draggedItemHeight = listView.getItemHeight(draggedItemPos!!)
                for (i in draggedItemPos!! until targetPosition) {
                    val nextHeight = listView.getItemHeight(i + 1)
                    listView.putItemHeight(i, nextHeight)
                }
                listView.putItemHeight(targetPosition, draggedItemHeight)
            } else if (step < 0) {
                val draggedItemHeight = listView.getItemHeight(draggedItemPos!!)
                for (i in draggedItemPos!! downTo targetPosition + 1) {
                    val nextHeight = listView.getItemHeight(i - 1)
                    listView.putItemHeight(i, nextHeight)
                }
                listView.putItemHeight(targetPosition, draggedItemHeight)
            }
            listView.adapter?.setDataSource(items)
            startTouchY += deltaH.toFloat()
            draggedItemViewTop = draggedItemViewTop!! +  deltaH
            draggedItemPos = targetPosition
            if (draggedItemView != null) {
                draggedItemView!!.visibility = View.VISIBLE
            }
            draggedItemView = listView.getItemView(draggedItemPos!!)
            if (draggedItemView != null) {
                draggedItemView!!.visibility = View.INVISIBLE
            }
        }
        listView.scrollHandler?.handleScrolling()
        listView.invalidate()
    }

    fun itemDraggingStopped() {
        if (draggedItemPos != null && draggedItemViewTop != null) {
            draggedItemPos = null

            hoverBitmapAnimation = hoverBitmap
            val scrollOffset = listView.scrollHandler!!.scrollOffset
            hoverBitmapBounds!!.offsetTo(0, draggedItemViewTop!! - (scrollOffset - scrollStart))
            val hoverBitmapBoundsCopy = Rect(hoverBitmapBounds)
            val draggedItemViewCopy = draggedItemView
            draggedItemView = null
            hoverBitmap = null
            draggedItemViewTop = null
            val hoverViewAnimator = ObjectAnimator.ofObject(
                hoverBitmapAnimation,
                "bounds",
                rectBoundsEvaluator,
                hoverBitmapBoundsCopy
            )
            hoverViewAnimator.addUpdateListener { _: ValueAnimator? -> listView.invalidate() }
            hoverViewAnimator.addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationStart(animation: Animator) {
                    listView.isEnabled = false
                }

                override fun onAnimationEnd(animation: Animator) {
                    if (draggedItemViewCopy != null) {
                        draggedItemViewCopy.visibility = View.VISIBLE
                    }
                    hoverBitmapAnimation = null
                    listView.isEnabled = true
                    listView.invalidate()
                }
            })
            hoverViewAnimator.start()
        } else {
            draggedItemPos = null
            draggedItemView = null
            draggedItemViewTop = null
        }
    }

    fun onItemMoveButtonPressed(
        position: Int,
        item: AbstractTreeItem?,
        itemView: View,
        touchX: Float,
        touchY: Float
    ) {
        startTouchY = itemView.top + touchY
        lastTouchY = startTouchY
        itemDraggingStarted(position, itemView)
    }

    fun onItemMoveButtonReleased(
        position: Int,
        item: AbstractTreeItem?,
        itemView: View?,
        touchX: Float,
        touchY: Float
    ) {
        itemDraggingStopped()
    }

    val isDragging: Boolean
        get() = draggedItemPos != null

    fun setDraggedItemView() {
        val draggedItemViewOld = draggedItemView
        draggedItemView = listView.getItemView(draggedItemPos!!)
        if (draggedItemView != null) {
            draggedItemView!!.visibility = View.INVISIBLE
        }
        if (draggedItemViewOld !== draggedItemView && draggedItemViewOld != null) {
            draggedItemViewOld.visibility = View.VISIBLE
        }
    }

    fun dispatchDraw(canvas: Canvas?) {
        if (hoverBitmap != null) {
            updateHoverBitmap()
            hoverBitmap!!.draw(canvas!!)
        }
        if (hoverBitmapAnimation != null) {
            hoverBitmapAnimation!!.draw(canvas!!)
        }
    }

    companion object {
        private val rectBoundsEvaluator: TypeEvaluator<Rect> = object : TypeEvaluator<Rect> {
            override fun evaluate(fraction: Float, startValue: Rect, endValue: Rect): Rect {
                return Rect(
                    interpolate(startValue.left, endValue.left, fraction),
                    interpolate(startValue.top, endValue.top, fraction),
                    interpolate(startValue.right, endValue.right, fraction),
                    interpolate(startValue.bottom, endValue.bottom, fraction)
                )
            }

            fun interpolate(start: Int, end: Int, fraction: Float): Int {
                return (start + fraction * (end - start)).toInt()
            }
        }
    }
}

fun getColor(color: Long): Int {
    // if alpha channel is not set - set it to max (opaque)
    if (color and -0x1000000 == 0L)
        return (color or -0x1000000).toInt()
    return color.toInt()
}
